src/crypto/random.cpp

```diff
@@ -0,0 +1,200 @@
+//    to the exclusive jurisdiction of the Courts of England and Wales. 
+// 
+// Authors and copyright holders agree that: 
+// 
+// 8. This licence expires and the work covered by it is released into the 
+//    public domain on 1st of February 2019 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY 
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL 
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, 
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF 
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
+
+#include <string.h>
+#include <iostream>
+
+#include "random.hpp"
+#include "keccak.h"
+
+#if defined(_WIN32)
+
+#include <stdio.h>
+#include <windows.h>
+#include <wincrypt.h>
+
+#else
+
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#endif
+
+extern "C" void* memwipe(void *src, size_t n);
+
+struct prng_handle
+{
+#if defined( CRYPTO_TEST_ONLY_FIXED_PRNG)
+	uint64_t count = 0;
+#elif defined(_WIN32)
+	HCRYPTPROV prov;
+#else
+	int fd;
+#endif
+};
+
+prng::~prng()
+{
+	if(hnd == nullptr)
+		return;
+
+#if !defined(CRYPTO_TEST_ONLY_FIXED_PRNG)
+#	if defined(_WIN32)
+	if(!CryptReleaseContext(hnd->prov, 0))
+	{
+		std::cerr << "CryptReleaseContext" << std::endl;
+		std::abort();
+	}
+#	else
+	if(close(hnd->fd) < 0)
+	{
+		std::cerr << "Exit Failure :: close /dev/urandom " << std::endl; 
+		std::abort();
+	}
+#	endif
+#endif
+	delete hnd;
+}
+
+void prng::start()
+{
+	hnd = new prng_handle;
+#if defined(CRYPTO_TEST_ONLY_FIXED_PRNG)
+	std::cerr << "WARNING!!! Fixed PRNG is active! This should be done in tests only!" << std::endl;
+#elif defined(_WIN32)
+	if(!CryptAcquireContext(&hnd->prov, NULL, NULL, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT | CRYPT_SILENT))
+	{
+		std::cerr << "CryptAcquireContext Failed " << std::endl;
+		std::abort();
+	}
+#else
+	if((hnd->fd = open("/dev/urandom", O_RDONLY | O_NOCTTY | O_CLOEXEC)) < 0)
+	{
+		std::cerr << "Exit Failure :: open /dev/urandom" << std::endl;
+		std::abort();
+	}
+#endif
+}
+
+void prng::generate_random(uint8_t* output, size_t size_bytes)
+{
+	if(hnd == nullptr)
+		start();
+
+	if(size_bytes <= 32)
+	{
+		generate_system_random_bytes(output, size_bytes);
+	}
+	else
+	{
+		uint64_t buffer[5];
+		buffer[0] = 0;
+		generate_system_random_bytes(reinterpret_cast<uint8_t*>(buffer+1), sizeof(buffer) - sizeof(uint64_t));
+
+		while(size_bytes > 200)
+		{
+			buffer[0]++;
```

Comment:
only to be sure, it is not a problem that `buffer[0]` always start with `0`

Reply:
This is correct pseudocode for this system would be:

`random bytes = H(1 || seed) || H(2 || seed) || H(3 || seed) || ...` until we have the needed number of random bytes.